src/my-bot.ts

```diff
@@ -1,38 +1,423 @@
+                    }
+                }
+                firebase.database().ref('matches/' + this.matchId.toString()).set(snapCopy);
+                if (snapCopy.hitmode) {
+                    target = this.track(gamestate,snapCopy);
+                } else if (snapCopy.hitmap) {
+                    target = this.randomShot(gamestate,snapCopy.hitmap);
+                } else {
+                    target = this.randomShot(gamestate);
+                }
+                return resolve(target);
+            })
+            .catch((err) => {console.log(err)});
+        })
+        
+        
+    }
+
+    private getNextTarget(position: Position): Position {
         var column = this.getNextColumn(position.Column);
         var row = column === 1 ? this.getNextRow(position.Row) : position.Row;
         return { Row: row, Column: column }
     }
 
-    private getNextRow(row) {
+    private getNextRow(row): string {
         var newRow = row.charCodeAt(0) + 1;
         if(newRow > 'J'.charCodeAt(0)) {
             return 'A';
         }
         return String.fromCharCode(newRow);
     }
 
-    private getNextColumn(column) {
+    private getNextColumn(column): number {
         return column % 10 + 1;
     }
+
+    private authenticate(): void {
+        firebase.auth().signInWithEmailAndPassword(
+            "david.may-miller@softwire.com",
+            "securePassword3"
+        );
+    }
+
+    private getShipPlace(shipPlaces: ShipPlace[], size: number): ShipPlace {
+        let counter:number = 0;
+        let place: ShipPlace = { StartingSquare: { Row: "A", Column: 1 }, EndingSquare : { Row: "A", Column: size } };
+        do {
+            let places: ShipPlace[] = shipPlaces.slice(0);
+            const row: string = String.fromCharCode(Math.floor(Math.random() * 10) + 65);
+            const column: number = Math.floor(Math.random() * 10) + 1;
+            if ((row.charCodeAt(0) > 75 - size) && (column > 11 - size)) {
+                continue;
+            }
+            if (((Math.random() < 0.5) && (row.charCodeAt(0) <= 75 - size)) || (column > 11 - size)) {
+                place = { StartingSquare: { Row: row, Column: column }, EndingSquare: { Row: String.fromCharCode(row.charCodeAt(0) + size - 1), Column: column } };
+            } else {
+                place = { StartingSquare: { Row: row, Column: column }, EndingSquare: { Row: row, Column: (column + size - 1) } };
+            }
+            places.push(place);
+            if (this.detectCollision(places)) {
+                counter++;
+            } else {
+                return place
+            }
+        } while(counter < 500)
+        return { StartingSquare: { Row: "A", Column: -1 }, EndingSquare : { Row: "A", Column: -1 } };
+    }
+
+    private detectCollision(shipPlaces: ShipPlace[]): boolean { //apart from the last one, none of these for loops should be more than 5 iterations so it's not as bad as it looks

```

Comment:
I like your naming - I don't actually think that your comment is needed, really. While there's a lot of nesting, this is cut down small enough that it's easy to pick up what it's doing.